### PR TITLE
Allow memento to send notifications to HTTP servers

### PR DESCRIPTION
**Pull request by mbt-msw**
Adds change notifications to memento.  When a notification URL is set, memento will make a POST request to that URL whenever a subscriber's call list is changed.  The POST body is a JSON document of the form:

```json
{ "impu": "<subscriber's impu>" }
```

This allows other servers in the cluster to react asynchronously to call lists changing.

### CODE REVIEW

File: include/httpnotifier.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+{
+public:
+  /// Constructor
+  HttpNotifier(HttpResolver* resolver);
+
+  /// Destructor
+  virtual ~HttpNotifier();
+
+  /// Set the URL to send notifications to.
+  void set_url(const std::string& notify_url);
```
Comment (AlexHockey) [inline include/httpnotifier.h]: Does the URL need to change after the object is created? Should this be part of the constructor?

File: include/mementoappserver.h [hunk 2]
```diff
@@ -77,6 +77,8 @@ class MementoAppServer : public AppServer
   /// @param  max_tokens             - Max number of tokens for the Cassandra load monitor (from configuration).
   /// @param  init_token_rate        - Initial token rate for the Cassandra load monitor (from configuration).
   /// @param  min_token_rate         - Minimum token rate for the Cassandra load monitor (from configuration).
+  /// @param  http_resolver          - HTTP resolver to use for HTTP connections.
+  /// @param  memento_notify_url     - HTTP URL that memento should notify when call lists change.
```
Comment (AlexHockey) [inline include/mementoappserver.h]: Presumably can be empty if you don't want to send a notify

File: src/httpnotifier.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  if (Utils::parse_http_url(notify_url, url_server, url_path) &&
+      !url_server.empty())
+  {
+    _http_connection = new HttpConnection(url_server,
+                                          true,
+                                          _http_resolver,
+                                          NULL,
+                                          NULL,
+                                          SASEvent::HttpLogLevel::PROTOCOL,
+                                          NULL);
```
Comment (AlexHockey) [inline src/httpnotifier.cpp]: FYI there's a shorter constructor that omits the stats table and the load monitor.

File: src/httpnotifier.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  notification.AddMember("impu", impu_value, notification.GetAllocator());
+
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  notification.Accept(writer);
+
+  std::map<std::string, std::string> headers;
+  std::string body = buffer.GetString();
+
+  HTTPCode http_code = _http_connection->send_post(
```
Comment (AlexHockey) [inline src/httpnotifier.cpp]: We normally align function parameters like this if they won't all fit on one line (e.g. using `cion=(0,g0)` in vim)

```
function(p1,
         p2);
```

File: src/ut/call_list_store_processor_test.cpp [hunk 5]
```diff
@@ -118,12 +123,14 @@ class CallListStoreProcessorWithLimitTest : public ::testing::Test
...
   }
 
   StrictMock<MockLoadMonitor> _load_monitor;
   CallListStoreProcessor* _clsp;
   MockCallListStore* _cls;
   LastValueCache* _stats_aggregator;
+  MockHttpNotifier* _http_notifier;
 };
 
 // Create a vector of call list store fragments that the mock
```
Comment (AlexHockey) [inline src/ut/call_list_store_processor_test.cpp]: Do you have any tests that check the notifier is actually called?

File: src/ut/httpnotifier_test.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+TEST_F(HttpNotifierTest, Notify)
+{
+  _http_notifier.set_url("http://notification.domain/notify");
+  bool ret = _http_notifier.send_notify("user@domain", 0);
+  EXPECT_TRUE(ret);
+  Request& req = fakecurl_requests["http://10.42.42.42:80/notify"];
+  EXPECT_EQ("POST", req._method);
+  EXPECT_EQ("{\"impu\":\"user@domain\"}", req._body);
+}
```
Comment (AlexHockey) [inline src/ut/httpnotifier_test.cpp]: If you need to keep `set_url`. you should also test what happens when it's not called.

File: src/ut/call_list_store_processor_test.cpp (hunk 5 above)
Comment (AlexHockey) [inline src/ut/call_list_store_processor_test.cpp]: Also do you have any tests that the notifier is not called if an empty/bad string is passed as the notify URL?

**Comment by AlexHockey** [comment]
I have reviewed. I think there are a few missing tests, and I'm not sure you need the `set_url` method. Other than that this looks fine. 

Please can you also update https://github.com/Metaswitch/memento/blob/dev/docs/memento_overview.md to describe the notification API?

**Comment by mbt-msw** [comment]
It turns out I'd mocked the HttpNotifier incorrectly, which is why my tests had missing parts.  I've also removed `set_url` as it is unnecessary, as you pointed out.

**Comment by AlexHockey** [comment]
Awesome sauce.
